compiler/qsc_rca/src/lib.rs

```diff
@@ -0,0 +1,841 @@
+    }
+
+    #[must_use]
+    pub fn get_expr(&self, id: ExprId) -> &ApplicationGeneratorSet {
+        self.exprs
+            .get(id)
+            .expect("expression compute properties not found")
+    }
+
+    #[must_use]
+    pub fn get_item(&self, id: LocalItemId) -> &ItemComputeProperties {
+        self.items
+            .get(id)
+            .expect("item compute properties not found")
+    }
+
+    #[must_use]
+    pub fn get_stmt(&self, id: StmtId) -> &ApplicationGeneratorSet {
+        self.stmts
+            .get(id)
+            .expect("statement compute properties not found")
+    }
+}
+
+/// The compute properties of an item.
+#[derive(Clone, Debug)]
+pub enum ItemComputeProperties {
+    /// The compute properties of a callable.
+    Callable(CallableComputeProperties),
+    /// The compute properties of a non-callable (for completeness only).
+    NonCallable,
+}
+
+impl Display for ItemComputeProperties {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        match &self {
+            ItemComputeProperties::Callable(callable_compute_properties) => {
+                write!(f, "Callable: {callable_compute_properties}")
+            }
+            ItemComputeProperties::NonCallable => write!(f, "NonCallable"),
+        }
+    }
+}
+
+/// The compute properties of a callable.
+#[derive(Clone, Debug)]
+pub struct CallableComputeProperties {
+    /// The application generator set for the callable's body.
+    pub body: ApplicationGeneratorSet,
+    /// The application generator set for the callable's adjoint specialization.
+    pub adj: Option<ApplicationGeneratorSet>,
+    /// The application generator set for the callable's controlled specialization.
+    pub ctl: Option<ApplicationGeneratorSet>,
+    /// The application generator set for the callable's controlled adjoint specialization.
+    pub ctl_adj: Option<ApplicationGeneratorSet>,
+}
+
+impl Display for CallableComputeProperties {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        let mut indent = set_indentation(indented(f), 0);
+        write!(indent, "CallableComputeProperties:",)?;
+        indent = set_indentation(indent, 1);
+        write!(indent, "\nbody: {}", self.body)?;
+        match &self.adj {
+            Some(spec) => write!(indent, "\nadj: {spec}")?,
+            None => write!(indent, "\nadj: <none>")?,
+        }
+        match &self.ctl {
+            Some(spec) => write!(indent, "\nctl: {spec}")?,
+            None => write!(indent, "\nctl: <none>")?,
+        }
+        match &self.ctl_adj {
+            Some(spec) => write!(indent, "\nctl-adj: {spec}")?,
+            None => write!(indent, "\nctl-adj: <none>")?,
+        }
+        Ok(())
+    }
+}
+
+/// A set of compute properties associated to a callable or one of its elements, from which the properties of any
+/// particular call application can be derived.
+#[derive(Clone, Debug)]
+pub struct ApplicationGeneratorSet {
```

Comment:
This is one of the names we wanted to explore alternatives for.

Reply:
Added it to #1285 to follow up on name alternatives.